### PR TITLE
Define themes in config.json

### PR DESCRIPTION
**Pull request by dtygel**
With this PR and also https://github.com/vector-im/riot-web/pull/3869 , themes aren't hardcoded in UserSettings anymore. Now any developper can create his/her themes and add them to config.json in riot-web. Important: the **first** theme declared in config.json will be the default for newcomers and users who didn't define their theme in personal settings.

### CODE REVIEW

File: src/components/structures/UserSettings.js [hunk 1]
```diff
@@ -97,27 +97,10 @@ const CRYPTO_SETTINGS_LABELS = [
...
-    {
-        id: 'theme',
-        label: 'Dark theme',
-        value: 'dark',
-    },
-];
+// Themes must be declared in config.json of riot-web. If they are not declared there, the FALLBACK_THEME will be used. Please **do not** include themes here, but directly in config.json! The default theme will always be the **first** theme in config.json.
 
+const FALLBACK_THEME = [{"label":"Light theme", "value": "light"}];
+const THEMES = [];
```
Comment (dbkr) [inline src/components/structures/UserSettings.js]: Since this is populated in `componentWillMount`, it should be a class variable.

**Comment by dbkr** [comment]
@ara4n do you want to take a look at this, as this is your department?

**Comment by ara4n** [comment]
the PR looks good, other than Dave's feedback - it's particularly important to handle fallback if the user selects a theme which a different riot instance might not support.

@dtygel - thanks for doing this; can you incorporate the feedback please?

**Comment by dtygel** [comment]
@ara4n and @dbkr : I've made changes to comply with Dave's warning about the need of a fallback theme when the user has an invalid theme. I had to make minor changes in riot-web as well, to create a way to validate themes and only allow valid themes to be presented to the user, besides having the fallback to the first valid theme in case the user has an invalid theme in his/her stored session data.

**Comment by dtygel** [comment]
Same here: Should I keep updating this PR to solve conflicts with new versions in develop? Or is it out of Riot's roadmap?

**Comment by dbkr** [comment]
Sorry for missing the updates on this, although I realise now that themes are basically compiled in to Riot (ie. the webpack build step builds the theme CSS and the main HTML file with a link element for each theme) so I don't think I really understand the benefit here. Anyone adding a theme will still need a Riot dev environment set up, they'll just have to add the theme in the config file rather than the Riot source. If anything I would have thought the list of themes should be auto-generated at build time?

**Comment by t3chguy** [comment]
There seem to be conflicting thoughts about this between @dbkr and @ara4n - https://github.com/vector-im/riot-web/pull/3869#issuecomment-355112734

**Comment by ara4n** [comment]
@dbkr: the benefit is surely to lock down the list of themes a given user is allowed to use on a given riot deployment.  E.g. status might not want the riot theme to be available on chat.status.im.

**Comment by dbkr** [comment]
Okay, yeah - that's useful, but that would just be a list of theme IDs to allow rather than moving the definitions for themes into the config, so this would be much simpler.

**Comment by dtygel** [comment]
I've been away or a while, and after rebasing, I see that many things changed related to themes. Maybe we should close these PRs. I don't see any sense to still try to follow up the development branch and rebase accordingly.

**Comment by jryans** [comment]
Thanks for making this contribution a while back. Since the code base has changed since this was opened, it no longer applies cleanly, and I don't think there's a need to keep it open in this state, as we can always find the code here again if needed. If you are still interested in pursuing this feature, please discuss with us in #riot-dev:matrix.org to find a good approach forward.

**Pull request closed without merging**